Images

niran opened the pull request:
This adds the ability to refer to thumbnail options using keys to a dictionary specified in the settings file instead of hard-coding the options in templates as suggested in the sorl documentation. It also establishes the `render_visual` convention intended for use in models that have a primary visual (image/video) associated with them.

niran committed:
Initial sketch of mixins to implement a sorl-backed image content item

niran committed:
Merge branch 'master' into images

niran committed:
Add lettuce

niran committed:
Add a test SorlImage and update the pip requirements

niran committed:
Enable the SorlImage admin for testing

niran committed:
Add image rendering spec

niran committed:
Rename render feature spec

niran committed:
Add a thumbnail presets spec

niran committed:
Add preset defaults scenarios

niran committed:
Merge remote-tracking branch 'main/master' into images

Conflicts:
	armstrong/core/arm_content/tests/arm_content_support/models.py

niran committed:
Add preset management functions

niran committed:
Add some thumbnail presets for manual testing

niran committed:
Fix broken function calls

niran committed:
Allow presets and defaults to be passed directly to functions that read those from the settings

niran committed:
Implement lettuce tests for presets

niran committed:
Implement lettuce tests for rendering

tswicegood commented on requirements/dev.txt:
```diff
@@ -4,8 +4,12 @@
 # that use case.
 -e git://github.com/bconstantin/django_polymorphic.git@2c47db8fcc284a92d2c9769ba503603fbea92660#egg=django_polymorphic
+
+PIL
```
FYI: you can do `-r ./sorl.txt` here to remove the duplication.

niran replied:
Didn't know that. Very handy.

tswicegood commented on armstrong/core/arm_content/utils/presets.py:
```diff
@@ -0,0 +1,16 @@
+
+
+def get_preset_args(preset_label, presets=None, defaults=None):
+    presets = presets or getattr(settings, 'ARMSTRONG_PRESETS', {})
```
I don't understand why this needs to be fetched from `settings` along with the `ARMSTRONG_PRESET_DEFAULTS` below on line 14.  I think it might be that "presets" is used twice...?

niran replied:
If my other comment doesn't adequately answer this, let me know.

tswicegood commented on armstrong/core/arm_content/utils/sorl.py:
```diff
@@ -0,0 +1,36 @@
+try:
+    from sorl.thumbnail import get_thumbnail
+
+    def get_preset_thumbnail(file_, preset_label, presets=None, defaults=None):
```
What happens if someone tries to run this an there's an `ImportError` because there's no sorl?

niran replied:
Good question. They'd get an `ImportError` anyway, but it'd be for `get_preset_thumbnail` instead of sorl (unless they did an `import *`, which would cause a `NameError` when `get_preset_thumbnail` is called), so it might be confusing. There may be a better way to handle this.

tswicegood replied:
I do like the idea of wrapping the import to the external library, but rather than `pass`ing.  Often a straight `ImportError` is not helpful.  What about raising an `ImproperlyConfigured` that references need to install sorl?

niran replied:
If we assume that this file will never be loaded unless sorl should be used, then that definitely works. I wasn't sure if that'd be the case.

niran replied:
In general, the non-sorl case hasn't been explored very much, but I think letting people who run into issues with it speak up is probably the most efficient approach.

tswicegood replied:
Yeah, let's go sorl with as much sane separation between `arm_content` and `sorl.*` calls as possible.  I think this gets us a good portion of the way there.

tswicegood commented on armstrong/core/arm_content/utils/presets.py:
```diff
@@ -0,0 +1,16 @@
+from django.core.exceptions import ImproperlyConfigured
+
+
+def get_preset_args(preset_label, presets=None, defaults=None):
```
In addition to the comments already made below, I would pass in `settings` as a kwarg so settings can be tested.

tswicegood commented on armstrong/core/arm_content/utils/sorl.py:
```diff
@@ -0,0 +1,36 @@
```
To stay consistent with the `videos` module in this component, this should probably be stored under `armstrong.core.arm_content.images.*`.

niran committed:
Work around fudge bug in Python 2.7.

https://bitbucket.org/kumar303/fudge/issue/12/use-as-a-context-manager-doesnt-work-in

niran committed:
Update requirements files to reflect the username change to armstrong

niran committed:
Reduce duplication in requirements by referencing other files

niran committed:
Be more explicit when choosing between kwargs and settings

niran committed:
Move utils to images

niran commented:
Made most of the suggested changes. Ready to be merged.

tswicegood commented:
Just did a quick pass through this.  The `sorl` import issues from my previous comments still need to be addressed.  Right now things can not work as expected and exceptions would be swallowed.  I also don't think that the sorl-specific backend should add features that aren't at least attempted in the other (see `SorlImageMixin.render_visual`).

For the record, I'm ok requiring sorl for images to work.  If we're going to go that route, though, we should require it and start raising exceptions if its not present.

niran committed:
Add a stub render_visual to BaseImageMixin

niran commented:
I agree that the current import exception situation is sub-optimal, but the obvious options for fixing it seem equally bad. I'll go ahead and remove the `try...except`s for the sake of better error messages until it causes a problem.

The sorl backend is the only backend right now, but I added a stub `render_visual` method to `BaseImageMixin` that raises `NotImplementedError`.

niran committed:
Remove try...except blocks around sorl imports.

This gives tests a hard dependency on sorl (if they didn't have one
before), but results in better error messages when sorl is missing or
outdated.

tswicegood commented on requirements/dev.txt:
```diff
@@ -1,11 +1,16 @@
 -e git://github.com/domain51/d51.django.virtualenv.test_runner#egg=d51.django.virtualenv.test_runner
 -e git://github.com/tswicegood/fabric#egg=fabric
--e git://github.com/armstrongcms/armstrong.dev#egg=armstrong.dev
+-e git://github.com/armstrong/armstrong.dev#egg=armstrong.dev
```
We should be relying on the released version instead of the dev version.

niran committed:
Switch to the released versions of armstrong packages

niran committed:
Add a method to get the URL of a thumbnail

niran committed:
Fix indentation

niran committed:
Allow the original image to be retrieved via the thumbnail functions

tswicegood commented on requirements/dev.txt:
```diff
@@ -1,11 +1,16 @@
+armstrong.dev
 coverage
 fudge
+lettuce
```
FYI, there's a bug in the released version of lettuce that prevents it from working if a `runserver` is currently open.  Actually hangs the system.  I'm updating it in my merge commit.
